src/document/doc-debug.js

```diff
@@ -35,100 +33,147 @@ function flattenDoc(doc) {
-      ")"
-    );
-  }
+    if (doc.type === "align") {
+      return doc.n === Number.NEGATIVE_INFINITY
+        ? "dedentToRoot(" + printDoc(doc.contents) + ")"
+        : doc.n < 0
+        ? "dedent(" + printDoc(doc.contents) + ")"
+        : doc.n.type === "root"
+        ? "markAsRoot(" + printDoc(doc.contents) + ")"
+        : "align(" +
+          JSON.stringify(doc.n) +
+          ", " +
+          printDoc(doc.contents) +
+          ")";
+    }
 
-  if (doc.type === "group") {
-    if (doc.expandedStates) {
+    if (doc.type === "if-break") {
       return (
-        "conditionalGroup(" +
-        "[" +
-        doc.expandedStates.map(printDoc).join(",") +
-        "])"
+        "ifBreak(" +
+        printDoc(doc.breakContents) +
+        (doc.flatContents ? ", " + printDoc(doc.flatContents) : "") +
+        (doc.groupId
+          ? (!doc.flatContents ? ', ""' : "") +
+            `, { groupId: ${JSON.stringify(printGroupId(doc.groupId))} }`
+          : "") +
+        ")"
       );
     }
 
-    return (
-      (doc.break ? "wrappedGroup" : "group") +
-      "(" +
-      printDoc(doc.contents) +
-      ")"
-    );
-  }
+    if (doc.type === "group") {
+      const optionsParts = [];
 
-  if (doc.type === "fill") {
-    return "fill" + "(" + doc.parts.map(printDoc).join(", ") + ")";
-  }
+      if (doc.break && doc.break !== "propagated") {
+        optionsParts.push("break: true");
+      }
 
-  if (doc.type === "line-suffix") {
-    return "lineSuffix(" + printDoc(doc.contents) + ")";
-  }
+      if (doc.id) {
+        optionsParts.push(`id: ${JSON.stringify(printGroupId(doc.id))}`);
+      }
+
+      const options =
+        optionsParts.length > 0 ? `, { ${optionsParts.join(", ")} }` : "";
+
+      if (doc.expandedStates) {
+        return `conditionalGroup([${doc.expandedStates
+          .map(printDoc)
+          .join(",")}]${options})`;
+      }
 
-  if (doc.type === "line-suffix-boundary") {
-    return "lineSuffixBoundary";
+      return `group(${printDoc(doc.contents)}${options})`;
+    }
+
+    if (doc.type === "fill") {
+      return `fill([${doc.parts.map(printDoc).join(", ")}])`;
+    }
+
+    if (doc.type === "line-suffix") {
+      return "lineSuffix(" + printDoc(doc.contents) + ")";
+    }
+
+    if (doc.type === "line-suffix-boundary") {
+      return "lineSuffixBoundary";
+    }
+
+    throw new Error("Unknown doc type " + doc.type);
   }
 
-  throw new Error("Unknown doc type " + doc.type);
+  function printGroupId(id) {
+    if (typeof id !== "symbol") {
+      return JSON.stringify(String(id));
+    }
+
+    if (id in symbolMap) {
+      return symbolMap[id];
+    }
+
+    const prefix = id.description || "symbol";
+    for (let counter = 0; ; counter++) {
+      const string = prefix + (counter > 0 ? ` #${counter}` : "");
+      if (!usedStrings.has(string)) {
+        usedStrings.add(string);
+        return (symbolMap[id] = `Symbol.for(${JSON.stringify(string)})`);
```

Comment:
I used like `return a = b` and `a = b = c`, long time no see...

Reply:
Aren't `[...usedStrings]` deepEqual `Object.keys(symbolMap)` ? I mean before you use `Symbol.for`.

So we can also use `string in symbolMap` as condition, right?

Reply:
No, both before and after it's `Object.value(symbolMap)`.

Reply:
Oh, right.

Reply:
SO, question is, is there a better way to check `Object.values(symbolMap).includes(string)`, I guess `Set` is a better choice.

Reply:
Better than what? I use a set already.

Reply:
JS doesn't have [bimaps](https://github.com/google/guava/wiki/NewCollectionTypesExplained#bimap)...

Reply:
I know, so you already have the best solution.

Reply:
Is in faster then `Set`? You can also use `Set` replace that `in`. If you add both key and value in it... Not sure about the performance.

Reply:
This code is a debugging tool :)

Reply:
Haha, but isn't it an interesting idea? 😄